### ISSUE TITLE
Train() not using all gpus

### ISSUE THREAD

**Issue by tesriram** [dmlc/xgboost#8021]
Hello. I am attempting to train a dataset for multiclass classification using 4 gpus. I set nworkers=4 when I create the cuda client and I believe this is working fine because all preprocessing (including creation of DaskQuantileDMatricises) uses all 4 GPUs. However when I call xgboost.dask.train and pass in the client, only 1 gpu is used. When I tried removing objective:softmax from Params, the model gets returned but does not train at all (booster['history'] is empty). I have tried a number of other things including setting n_jobs n_threads and/or n_gpus but nothing seems to work. Any help would be appreciated. Thank you

**Comment by trivialfis** [comment]
Could you please share a reproducible example?

**Comment by tesriram** [comment]
Unfortunately I cannot post screenshots of the code. From what I understand, the important pieces of code are: 
With localCUDACluster(nworkers=4) as cluster:
   With Client(cluster) as client:
       Main(client)

In Main:
*Read in dataset and perform data cleaning that gives datasets as cudf dataframes
Xy = xgboost.dask.DaskDeviceQuantileDMatrix(client, dask_cudf.from_cudf(xtrain, npartitions=-1), dask_cudf.from_cudf(ytrain, npartitions=-1))
Booster = xgboost.dask.train(client,
{'objective' : 'multi:softmax,
'eval_error : 'merror',
'tree_method : 'gpu_hist,
'num_classes : length(<class clist>)},
Xy)

I tried to follow online examples as closely as possible. Is there anything I am missing or anything else I can try?

**Comment by trivialfis** [comment]
I think the correct parameter for `LocalCUDACluster` is `n_workers` instead of `nworkers`.

**Comment by tesriram** [comment]
That was a typo. I do use n_workers and all 4 GPUs are being used up until the train function. I'm wondering if multi-gpu training is even supported for xgboost multiclass classification?

**Comment by trivialfis** [comment]
It's supported and I run it from time to time. That's why I need access to a reproducible example. I can't guess what's happening on your end.

**Comment by tesriram** [comment]
My work does not permit me to directly share code. The closest I can give is that I essentially copied the code from this git repo in demo/dask/gpu_training.py, but with n_workers = 4 and did not set threads_per_worker.
My two guesses here are:
1. The objective parameter set to multi:softmax/softprob does not allow multi gpu training (not sure if this is what you used)
2. The output we get from DaskDeviceQuantileMatrox cannot be parsed using multi gpu (the input to this function are dask_cudf dataframe and series)

I apologize that I cannot share my code and your help with my issue is much appreciate

**Comment by trivialfis** [comment]
I use both of them. My guess is that the data is not well balanced so some GPUs appear idle.

**Comment by trivialfis** [comment]
A little bit more context, XGBoost runs on the workers that own the data. So if the data is concentrated in one worker then XGBoost will launch a single task on that worker with all the other workers ignored.

**Comment by tesriram** [comment]
Ok turns out npartitions was 1 the whole time. I've changed it to npartitions = 4. However now I'm receiving a caught signal 7 (bus error: non existent physical address). In Main I have a line: client.run('set_allocator', 'managed') which I thought would manage all gpu memory for me. Any ideas as to why I'm still receiving this error and/or how I can fix it?

**Comment by trivialfis** [comment]
Haven't used `client.run('set_allocator', 'managed')` before, I don't think it's necessary. If you want to have a memory pool, you can take a look into using RMM: https://github.com/dmlc/xgboost/tree/master/demo/rmm_plugin

**Comment by tesriram** [comment]
Creating new issue for this
